daemons/attrd/attrd_sync.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2022 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <crm/msg_xml.h>
+#include <crm/common/attrd_internal.h>
+
+#include "pacemaker-attrd.h"
+
+static GHashTable *waitlist = NULL;
+
+struct waitlist_node {
+    enum attrd_sync_point sync_point;
+
+    char *client_id;
+    uint32_t ipc_id;
+    uint32_t flags;
+
+    char *attr_name;
+};
+
+static gboolean
+waitlist_node_GEqualFunc(gconstpointer a, gconstpointer b)
+{
+    const struct waitlist_node *a_node = a;
+    const struct waitlist_node *b_node = b;
+
+    return pcmk__str_eq(a_node->client_id, b_node->client_id, pcmk__str_none) &&
+           pcmk__str_eq(a_node->attr_name, b_node->attr_name, pcmk__str_none);
+}
+
+static void
+free_waitlist_node(gpointer data)
+{
+    struct waitlist_node *wl = (struct waitlist_node *) data;
+
+    free(wl->client_id);
+    free(wl->attr_name);
+    free(wl);
+}
+
+static const char *
+sync_point_str(enum attrd_sync_point sync_point)
+{
+    if (sync_point == attrd_sync_point_local) {
+        return "local";
+    } else if  (sync_point == attrd_sync_point_cluster) {
+        return "all nodes";
+    } else {
+        return "unknown";
+    }
+}
+
+void
+attrd_add_client_to_waitlist(pcmk__request_t *request)
+{
+    const char *sync_point = crm_element_value(request->xml, PCMK__XA_ATTR_SYNC_POINT);
+    const char *attr_name = crm_element_value(request->xml, PCMK__XA_ATTR_NAME);
+    struct waitlist_node *wl = NULL;
+    int callid;
+
+    if (sync_point == NULL) {
+        return;
+    }
+
+    if (waitlist == NULL) {
+        waitlist = g_hash_table_new_full(g_int_hash, waitlist_node_GEqualFunc,
+                                         NULL, free_waitlist_node);
+    }
```

Comment:
For completeness we should free the table at exit

Reply:
Fixed.